api-get.go

```diff
@@ -286,38 +330,105 @@ func (r *objectReadAtCloser) ReadAt(b []byte, offset int64) (int, error) {
 	if dataMsg.Error == nil {
 		// If offset+bytes read is equal to objectSize
 		// we have reached end of file, we return io.EOF.
-		if offset+bytesRead == r.objectSize {
+		if offset+bytesRead == o.objectInfo.Size {
 			return dataMsg.Size, io.EOF
 		}
 		return dataMsg.Size, nil
 	}
 
 	// Save any error.
-	r.prevErr = dataMsg.Error
+	o.prevErr = dataMsg.Error
 	return dataMsg.Size, dataMsg.Error
 }
 
-// Closer is the interface that wraps the basic Close method.
+// Seek sets the offset for the next Read or Write to offset,
+// interpreted according to whence: 0 means relative to the
+// origin of the file, 1 means relative to the current offset,
+// and 2 means relative to the end.
+// Seek returns the new offset and an error, if any.
 //
-// The behavior of Close after the first call returns error for
+// Seeking to a negative offset is an error. Seeking to any
+// positive offset is legal, subsequent io operations succeed until the
+// underlying object is not closed.
+func (o *Object) Seek(offset int64, whence int) (n int64, err error) {
+	if o == nil {
+		return 0, ErrInvalidArgument("Object is nil")
+	}
+
+	// Locking.
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	if o.prevErr != nil {
+		// At EOF seeking is legal, for any other errors we return.
+		if o.prevErr != io.EOF {
+			return 0, o.prevErr
+		}
+	}
+
+	// Negative offset is valid for whence of '2'.
+	if offset < 0 && whence != 2 {
+		return 0, ErrInvalidArgument(fmt.Sprintf("Object: negative position not allowed for %d.", whence))
+	}
+	switch whence {
+	default:
+		return 0, ErrInvalidArgument(fmt.Sprintf("Object: invalid whence %d", whence))
+	case 0:
+		if offset > o.objectInfo.Size {
+			return 0, io.EOF
+		}
+		o.currentOffset = offset
+	case 1:
+		if o.currentOffset+offset > o.objectInfo.Size {
+			return 0, io.EOF
+		}
+		o.currentOffset += offset
+	case 2:
+		if o.currentOffset+offset > o.objectInfo.Size {
```

Comment:
> this check doesnt make sense. We need to check "offset > 0".

offset can be greater than 0 its a valid case, what isn't valid is the effective offset being negative.